docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,114 @@
+# Initializers in if statements
+
+# Summary
+
+Introduce an initializer expression that declares and initializes variables in if statements.
+
+# Motivation
+
+Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's block simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the condition's pertaining block, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
+
+The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope.
+
+# Design
+
+If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. The variables declared by the initializer are only available to the block of that condition and will be undefined to the `elseif` conditions and blocks as well as the `else` block; any code after the if statement won't have the variables defined either.
+
+In the former case, the value of the first declared variable will be checked.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() then
+   print(b)
+else
+   print("false")
+end
+```
+
+`Output: true`
+
+In the latter case, the `exp` condition is checked rather than the initializer.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() where b == false then
+   print(b)
+else
+   print("true")
+end
```

Comment:
This actually gave me a motivating reason why it might be useful for `b` to be visible in the `else` branch. Say you have some animal:

```lua
type Dog = { tag: "dog", pet: true, name: string }
type Cat = { tag: "cat", pet: true, name: string }
type Sheep = { tag: "sheep", pet: false }
type Animal = Dog | Cat | Sheep

if local animal = getAnimal() where animal.pet then
  print(`Hello {animal.name}!`)
else
  print(`{animal.tag} is not a pet`) -- generalizes to all current and future `animal.tag`s
end
```

Reply:
Another case:
```
if local success, msg = pcall(f) where success then
    print(`all went well`)
else
    print(`f failed with: {msg}`)
end
```

Reply:
> This actually gave me a motivating reason why it might be useful for `b` to be visible in the `else` branch. Say you have some animal:
> 
> ```lua
> type Dog = { tag: "dog", pet: true, name: string }
> type Cat = { tag: "cat", pet: true, name: string }
> type Sheep = { tag: "sheep", pet: false }
> type Animal = Dog | Cat | Sheep
> 
> if local animal = getAnimal() where animal.pet then
>   print(`Hello {animal.name}!`)
> else
>   print(`{animal.tag} is not a pet`) -- generalizes to all current and future `animal.tag`s
> end
> ```

> Another case:
> 
> ```
> if local success, msg = pcall(f) where success then
>     print(`all went well`)
> else
>     print(`f failed with: {msg}`)
> end
> ```

Seeing it now allowing `b` to be visible in the `else` branch (and most likely `elseif` branches) would be useful.

Reply:
IMO `pcall` is the most compelling reason to scope the locals this way, but this scoping behavior is a consistent footgun in C++ where bindings in conditions are scoped to the whole if statement and you can trivially end up dereferencing null pointers accidentally using them in the wrong branch. I think, outside of `pcall`, scoping the bindings to the whole block means making more bugs with accidentally using `nil` possible, and I see little benefit since you can already write the following to have this scoping behavior.
```lua
do local success, msg = pcall(f)
    if success then
        print(`all went well`)
    else
        print(`f failed with: {msg}`)
    end
end
```